classes/Village.php

```diff
@@ -38,6 +58,34 @@ public function __construct($system, $village = '', array $village_row = []) {
             $this->coords = VillageManager::getLocation($this->system, $this->village_id);
             $this->relations = VillageManager::getRelationsForVillage($this->system, $this->village_id);
             $this->policy = new VillagePolicy($this->policy_id);
+            $this->buildings = VillageUpgradeManager::getBuildingsForVillage($this->system, $this->village_id);
+            $this->upgrades = VillageUpgradeManager::getUpgradesForVillage($this->system, $this->village_id);
+            $this->active_upgrade_effects = VillageUpgradeManager::initializeEffectsForVillage($this->system, $this->upgrades);
+            $this->construction_speed += ($this->active_upgrade_effects[VillageUpgradeConfig::UPGRADE_EFFECT_CONSTRUCTION_SPEED] / 100);
+            $this->research_speed += ($this->active_upgrade_effects[VillageUpgradeConfig::UPGRADE_EFFECT_RESEARCH_SPEED] / 100);
+            $this->resource_capacity += $this->active_upgrade_effects[VillageUpgradeConfig::UPGRADE_EFFECT_RESOURCE_CAPACITY];
+            if ($this->policy->construction_speed > 0) {
+                $result = $system->db->query("SELECT `village_id`, `construction_score` FROM `villages`");
+                $village_scores = $system->db->fetch_all($result);
+                $num_villages = 0;
+                foreach ($village_scores as $score) {
+                    if ($score['village_id'] != $this->village_id && $score['construction_score'] > $this->construction_score) {
+                        $num_villages++;
+                    }
+                }
+                $this->construction_speed += ($num_villages * $this->policy->construction_speed) / 100;
+            }
```

Comment:
Consider extracting this to a method and then doing something like `$this->construction_speed = $this->getConstructionSpeed()`